jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/messages/AbstractMessageSink.java

```diff
@@ -20,12 +20,35 @@
 
 public abstract class AbstractMessageSink implements MessageSink
 {
-    protected final CoreSession session;
-    protected final MethodHandle methodHandle;
+    private final CoreSession session;
+    private final MethodHandle methodHandle;
+    private final boolean autoDemand;
 
-    public AbstractMessageSink(CoreSession session, MethodHandle methodHandle)
+    public AbstractMessageSink(CoreSession session, MethodHandle methodHandle, boolean autoDemand)
```

Comment:
javadoc is needed.

I really question the need for autoDemand=false, as if it is false, how does the user know how many frames to demand? how do they know to demand more when  non-fin frame has arrived and they are waiting for the full message?

Perhaps there should be a demand method on this class, that demand(1) means 1 message and the impl will do frame demands until the fin frame is received?

Reply:
Depends on where the `MessageSink` is used.
Currently, it is not meant to be used in WebSocket core, but if it is, the semantic is that of `CoreSession.demand(long)`, so asking for frames.
Typically `demand(1)` from the application is enough, because then the `MessageSink` will take over the demand when it is not invoking the application.
So, depending on the `MessageSink` implementation, `demand(1)` may mean "next frame" or "next message".

For higher levels, either there is no demand API (Jakarta), so the `MessageSink` is always auto-demanding, or there is `Session.demand()` (with no parameters), so again the "demand" meaning depends on the `MessageSink` implementation.

Reply:
Then the javadoc here needs to explain that `autoDemand==false` does not mean the `MessageSink` will never issue demand, just that it will never issue unprovoked demand.   With `autoDemand==false` a sink may still issue demand to complete a partial message or to continue a stream.  But it will not demand if idle.

Perhaps we should call it `idleDemand` here rather than `autoDemand`?

Reply:
I have updated the javadocs, and kept `autoDemand` -- don't think `idleDemand` is right.

Reply:
It is better, but a bit wordy and not exactly clear.  Also no documentation on the actual constructor.   How about:
```java
/**
 * @param autoDemand If true, the user of the MessageSink need not (and must not) use the 
 *                                      CoreSession.demand method. The MessageSink implementation will manage 
 *                                      demand so that a single message will be delivered at once and frames for the next
 *                                      message will only be demanded after completion of the handling of the previous
 *                                      message.
 *                                      If false, the user of the MessageSink must use the CoreSession.demand 
 *                                      method to initiate handling of a new message. The MessageSink implementation
 *                                      must manage demand itself only when non final frames are received and 
 *                                      partial messages are not being delivered to the user of the MessageSink.
 */

Reply:
There is no "user of the MessageSink", so it's still not clear.
And there are too many assumptions about what the `MessageSink` does (not all of them accumulate into messages).

What do you find not clear in the current version?

Reply:
If there is no "user" then who calls demand when not in auto demand mode?  If the class is not used then delete it!

Reply:
@sbordet what I find unclear about the current is:
* the javadoc is on the class and not on the parameter in the constructor.
*  it is over 4 paragraphs with references to "the former case" and "the later case" indicating structural complexity
* what does it mean for a methodHandle to explicitly demand? A method handle just invokes another method, which itself might demand or delegate the demand to another method or another process.   Demand is not managed by the method handle, it is managed by something using the MessageSink.

There are two parties to this relationship: the application as represented by the methodHandler; the implementation of the MessageSink.  Let's just call them "application" (what I called user) and ("messageSink").

There are two modes: autoDemand==true/false.    So we just need to be crystal clear about what the two parties can/can't/should do in each mode

autoDemand==true : The application need not and must not call demand.  The messageSink will manage all demand to a single stream of messages.

autoDemand==false:  The application need to call demand to commence reading of a message.  However the messageSink will then manage demand, if need be, so that sufficient frames can be received to invoke the application.  If  the messageSink is delivering complete message to the application, then it will manage demand when non final frames are received.  Once a message (partial or complete) has been delivered to the application, then the application must use demand in order to receive the next message (partial or complete)

Reply:
> it is over 4 paragraphs with references to "the former case" and "the later case" indicating structural complexity

It's a boolean parameter, so I need to enumerate 2 cases. What structural complexity are you talking about?

> what does it mean for a methodHandle to explicitly demand?

Nowhere in the javadoc is written that "MethodHandle to explicitly demand".

Application writers will never look at these javadocs, they are not even aware `MessageSink` exists.
These javadocs are for implementers of `MessageSink`, so IMHO it is ok to explain in the class what `autoDemand` is rather than having many paragraphs of explanations in the constructor parameter, to be repeated in the getter and in `autoDemand()`.

Part of what you explain above may be added to documentation for application writers, but it is irrelevant and incomplete for `MessageSink` implementers.

The current javadocs are explicitly tailored for `MessageSink` implementers about how to write a `MessageSink` implementation correctly not about what applications can or cannot do.

Reply:
The text is structurally complex.  When reading one paragraph and it says "the former case", you have to look back at the previous paragraph and see what that is.    Just enumerate the two cases rather than spread it out over multiple paragraphs.

Your current text says: "The application function passed to the constructor as a MethodHandle can either explicitly demand for more WebSocket frames ...",  so yeah technically is says the application function is demanding, but it's hard to read and overly complex structure.

At the end, I'm just saying that I think I well understand how demand is meant to work for message sinks in both modes, but that I find the javadoc does not well express that, nor is it in the right place.   I'm constantly re-reading it, getting lost/confused, re-reading the bit I just read before.  It does not express what I understand the two modes to be.

If you don't like my suggestions, then at the very least can you:
 + at least have some javadoc to the parameter in the constructor that has a proper link to a section that clearly describes autoDemand
 + explain it in two cases: what true means and what false means, without the need to refer back and forth between the paragraphs.
 + your example is good, but it is unclear which mode (or both) it relates to. Perhaps have examples for both the true and false case.

Reply:
@gregw updated javadocs.

Reply:
Resovling this big conversation in favour of the micro discussions above.